File: leetcode/tree/binaryTreePaths.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    def __str__(self):
        return  str(self.val)
class Solution(object):
    result = []


    def binaryTreePaths(self, root):
            # Write your code here
        path = ''
        res = []
        self.TreePathsHelper(root, path, res)
        return res

    def TreePathsHelper(self, root, path, res):
        if root==None:
            return #给头是空设计的
        path+=str(root.val)
        if root.left:
            self.TreePathsHelper(root.left,path+"->",res)
        if root.right:
            self.TreePathsHelper(root.right,path+"->",res)
        if root.right is None and root.left is None:
            res.append(path)


if __name__ == '__main__':
    a = TreeNode(1)
    b = TreeNode(2)
    c = TreeNode(2)
    d=TreeNode(3)
    e=TreeNode(5)
    h=TreeNode(4)
    g = TreeNode(7)
    a.left = b
    a.right = c
    c.left=d
    b.left=e
    b.right=h
    e.left=g

    print(Solution().binaryTreePaths(a))

File: java/src/Tree/BinaryTreePaths.java
package Tree;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class BinaryTreePaths {
    private List<String> res;
    public List<String> binaryTreePaths(TreeNode root) {
        String pre = "";
        res = new ArrayList<>();
        if(root == null){
            return res;
        }
        getPath(root,pre);
        return res;
    }

    public void getPath(TreeNode root,String pre){
        if(root == null){
            return;
        }
        if(root.left==null&&root.right==null){
            System.out.println(root);
            String del = pre+"->"+root.val;
            res.add(del.substring(2));
        }
        else{
            getPath(root.left,pre+"->"+root.val);
            getPath(root.right,pre+"->"+root.val);
        }

    }
}
